drivers/hmis/app/graphql/mutations/delete_form_definition.rb

```diff
@@ -17,7 +17,16 @@ def resolve(id:)
       raise 'not found' unless definition
       raise 'can only delete draft forms' unless definition.draft?
 
-      definition.destroy!
+      has_other_versions = definition.all_versions.count > 1
+
+      Hmis::Form::Definition.transaction do
+        # If this is the last remaining version of this form, remove all Form Instances.
+        definition.instances.each(&:destroy!) unless has_other_versions
+
+        # Soft-delete the Form Definition. This will error if there are any form_processors or external_form_submissions
+        # that reference this particular form, which there shouldn't be because it's a Draft.
+        definition.destroy!
+      end
```

Comment:
Looks like there is a race condition here. Maybe not worth the trouble since it's unlikely to be triggered, but could fix it by wrapping it in an advisory lock. 

Normally you could use definition.with_lock {} but the logic here involves multiple FD rows

```ruby
      Hmis::Form::Definition.with_advisory_lock("hmis-fd-delete-mutation") do
        has_other_versions = definition.all_versions.count > 1
        Hmis::Form::Definition.transaction do
          # If this is the last remaining version of this form, remove all Form Instances.
          definition.instances.each(&:destroy!) unless has_other_versions

          # Soft-delete the Form Definition. This will error if there are any form_processors or external_form_submissions
          # that reference this particular form, which there shouldn't be because it's a Draft.
          definition.destroy!
        end
      end
```

Reply:
Huh I don't see where the race condition is, can you point it out?

Reply:
I'm going to merge so I can get this up on QA, but will open a second PR if needed (And would love to see what I'm missing!)

Reply:
> I'm going to merge so I can get this up on QA, but will open a second PR if needed (And would love to see what I'm missing!)

Sounds good! So I think it's around the conditional delete on has_other_versions conditional behavior. In this case this condition is unlikely to occur, maybe even vanishingly unlikely so perhaps this is just academic.

But Let's say we receive two simultaneous requests and they are processed in parallel to delete two different FDs records with the same identifier. Then let's say thread 1 queries for has_other_versions, then thread 2 queries has_other_versions at the same, before any delete occurs. They both get a count  of 2 so neither of them destroys instances. But if they ran in series the second request would see a count of 1 and the delete would happen. I guess the outcome is we end up with orphan form instances which maybe doesn't matter than much.

Reply:
Ahh I see, thanks for the explanation. I think we can ignore it, since deleting a form draft is an uncommon action, and the negative outcome (orphaned instances) is not that bad.